### PR TITLE
Backup/Restore commands

### LINKED ISSUE #1292: Migrate backup/restore from Gogs
https://github.com/gogits/gogs/commit/ca2cfaf71e33f42c00f5baf4792ba8396e6d0042

PR https://github.com/go-gitea/gitea/pull/1637

### PR DESCRIPTION
**Pull request by bkcsoft**
Backup/Restore commands! 🎉 

```
NAME:                                                                                                                                                                                                                                                                           
   gitea backup - Backup files and database                                                                                                                                                                                                                                     

DESCRIPTION:
   Backup dumps and compresses all related files and database into zip file,
   which can be used for migrating Gitea to another server. The output format is meant to be
   portable among all supported database engines.

OPTIONS:
   --config FILE, -c FILE   Custom configuration FILE path (default: "custom/conf/app.ini")
   --tempdir PATH, -t PATH  Temporary directory PATH (default: "/tmp")
   --target PATH            Target directory PATH to save backup archive (default: "./")
   --verbose, -v            Show process details
   --db                     Backup the database (default: true)
   --repos                  Backup repositories (default: true)
   --data                   Backup attachments and avatars (default: true)
   --custom                 Backup custom files (default: true)
```

```
NAME:
   gitea restore - TBD
```

TODO-list:
- [x] Migrate commit
- [ ] Add documentation
  - [ ] Add to features list: https://github.com/go-gitea/docs/blob/master/content/page/index.en-us.md
- [x] Add deprecation notice to  `gitea dump`
- [x] Add Changelog with deprecation notice

Closes #1292

### CODE REVIEW

**Comment by bkcsoft** [comment]
Currently seg-faults on dumping database. @lunny do you know why? 🙂 

```
$ ./gitea backup                                                                                                                                                                                                                                                 
panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                                                                         
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x849367]

code.gitea.io/gitea/modules/log.(*XORMLogBridge).ShowSQL(0x0, 0xc420188475, 0x1, 0x1)
        /home/bkc/sources/gopath/src/code.gitea.io/gitea/modules/log/xorm.go:134 +0x17
code.gitea.io/gitea/vendor/github.com/go-xorm/xorm.(*Engine).ShowSQL(0xc4201e8960, 0xc420188475, 0x1, 0x1)
        /home/bkc/sources/gopath/src/code.gitea.io/gitea/vendor/github.com/go-xorm/xorm/engine.go:53 +0x52
code.gitea.io/gitea/models.SetEngine(0xc420400140, 0x13351f4)
        /home/bkc/sources/gopath/src/code.gitea.io/gitea/models/models.go:265 +0x1c5
code.gitea.io/gitea/cmd.runBackup(0xc420400140, 0x0, 0x0)
        /home/bkc/sources/gopath/src/code.gitea.io/gitea/cmd/backup.go:82 +0xdb
code.gitea.io/gitea/vendor/github.com/urfave/cli.HandleAction(0x11d88a0, 0x1390aa8, 0xc420400140, 0xc420072000, 0x0)                                                                                                                                                            
        /home/bkc/sources/gopath/src/code.gitea.io/gitea/vendor/github.com/urfave/cli/app.go:471 +0xb9                                                                                                                                                                          
code.gitea.io/gitea/vendor/github.com/urfave/cli.Command.Run(0x1334fd2, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x135e236, 0x19, 0x0, ...)                                                                                                                                                
        /home/bkc/sources/gopath/src/code.gitea.io/gitea/vendor/github.com/urfave/cli/command.go:191 +0xb4b
code.gitea.io/gitea/vendor/github.com/urfave/cli.(*App).Run(0xc4200016c0, 0xc42008a060, 0x2, 0x2, 0x0, 0x0)                                                                                                                                                                     
        /home/bkc/sources/gopath/src/code.gitea.io/gitea/vendor/github.com/urfave/cli/app.go:241 +0x65f                                                                                                                                                                         
main.main()
        /home/bkc/sources/gopath/src/code.gitea.io/gitea/main.go:46 +0x455   
```

File: cmd/restore.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	if c.IsSet("config") {
+		setting.CustomConf = c.String("config")
+	} else if !com.IsExist(configFile) {
+		log.Fatal(0, "'--config' is not specified and custom config file is not found in backup")
+	} else {
+		setting.CustomConf = configFile
+	}
+	setting.NewContext()
+	models.LoadConfigs()
+	models.SetEngine()
```
Comment (lunny) [inline cmd/restore.go:116]: handle error

**Comment by lunny** [comment]
Any update @bkcsoft

**Comment by bkcsoft** [comment]
@lunny I'll see if I have time to look at this next week

**Comment by hoonetorg** [comment]
Regarding https://github.com/go-gitea/gitea/issues/2682: 
When there will be a "gitea restore" command, there should be no reason to use zip iiac. 
Also tgz can be handled fine in windows with a lot of programs afaik.

**Comment by bkcsoft** [comment]
Yeah, I agree that when we have `gitea restore` we could go for compressed tar-archives instead. Esp. since it's builtin to the language (instead of yet another dependency) https://golang.org/pkg/archive/tar/

**Comment by elvarb** [comment]
Will this be completed soon?

**Comment by lafriks** [comment]
@elvarb there is now other PR #2917

**Comment by elvarb** [comment]
@lafriks ok, glad to see lots of activity :)

**Comment by appleboy** [comment]
@bkcsoft conflicts.

**Comment by bkcsoft** [comment]
Closing this in favour of #2917

**Pull request closed without merging**